### PR TITLE
Update API Extractor to support TypeScript 5.3.3

### PR DESCRIPTION
**Pull request by octogonz**
## Summary

Fixes https://github.com/microsoft/rushstack/issues/4404

## Details

~👉 This is a draft PR because TypeScript 5.3 has not shipped yet. We generally would not do an official API Extractor release that bundles a beta version of the TypeScript compiler.~

## How it was tested

Confirmed that the API Extractor tests pass.

### CODE REVIEW

File: apps/api-extractor/src/analyzer/TypeScriptInternals.ts [hunk 1]
```diff
@@ -82,14 +82,19 @@ export class TypeScriptInternals {
...
+    program: ts.Program,
     sourceFile: ts.SourceFile,
     moduleNameText: string,
     mode: ts.ModuleKind.CommonJS | ts.ModuleKind.ESNext | undefined
   ): ts.ResolvedModuleFull | undefined {
     // Compiler internal:
-    // https://github.com/microsoft/TypeScript/blob/v4.7.2/src/compiler/utilities.ts#L161
-
-    return (ts as any).getResolvedModule(sourceFile, moduleNameText, mode);
+    // https://github.com/microsoft/TypeScript/blob/v5.3-beta/src/compiler/types.ts#L4678
```
Comment (jakebailey) [inline apps/api-extractor/src/analyzer/TypeScriptInternals.ts:91]: For backwards compat, you'll probably want to still try the old function for versions before 5.3

Comment (octogonz) [reply apps/api-extractor/src/analyzer/TypeScriptInternals.ts:91]: API Extractor bundles a specific version of TypeScript; it does not try to interoperate with arbitrary versions.

Comment (jakebailey) [reply apps/api-extractor/src/analyzer/TypeScriptInternals.ts:91]: Oh, gotcha; had assumed differently a la heft.

**Comment by rvion** [comment]
any way to help supporting more recent version of typescript for API Extractor ? not having `as const` nor proper branch subtyping is really forcing more and more antipatterns in code to keep using api-extractor

File: apps/api-extractor/package.json [hunk 2]
```diff
@@ -48,7 +48,7 @@
     "resolve": "~1.22.1",
     "semver": "~7.5.4",
     "source-map": "~0.6.1",
-    "typescript": "~5.0.4"
+    "typescript": "5.3.0-beta"
```
Comment (D4N14L) [inline apps/api-extractor/package.json:51]: Should probably target 5.3.3

File: apps/api-extractor/src/analyzer/TypeScriptInternals.ts (hunk 1 above)
Comment (D4N14L) [inline apps/api-extractor/src/analyzer/TypeScriptInternals.ts:91]: Update the link to point to the non-beta

File: common/changes/@microsoft/api-extractor/octogonz-ae-ts-5.3.0-beta_2023-10-24-21-05.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/api-extractor",
+      "comment": "Update API Extractor to support TypeScript 5.3.0-beta",
```
Comment (D4N14L) [inline common/changes/@microsoft/api-extractor/octogonz-ae-ts-5.3.0-beta_2023-10-24-21-05.json:5]: Update

File: common/config/rush/common-versions.json [hunk 4]
```diff
@@ -76,7 +76,9 @@
       // For testing Heft with TS V3
       "~3.9.10",
       // For testing Heft with TS V4
-      "~4.9.5"
+      "~4.9.5",
+
+      "5.3.0-beta"
```
Comment (D4N14L) [inline common/config/rush/common-versions.json:81]: Update

**Review by D4N14L — APPROVED** [review]

File: apps/api-extractor/src/analyzer/TypeScriptInternals.ts (hunk 1 above)
Comment (octogonz) [reply apps/api-extractor/src/analyzer/TypeScriptInternals.ts:91]: done

**Pull request merged**